apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryResults.cs

```diff
@@ -122,10 +163,32 @@ public BigQueryResults(BigQueryClient client, GetQueryResultsResponse response,
         public BigQueryPage ReadPage(int pageSize)
         {
             GaxPreconditions.CheckArgumentRange(pageSize, nameof(pageSize), 1, int.MaxValue);
-            // Make sure we start off by trying to read the right page size...
-            var options = GetOptionsWithPageSize(pageSize);
-            var page = _client.ListRows(TableReference, Schema, options).ReadPage(pageSize);
-            return new BigQueryPage(page, Schema, JobReference, TableReference);
+            GetQueryResultsOptions clonedOptions = _options?.Clone() ?? new GetQueryResultsOptions();
+            List<BigQueryRow> rows = new List<BigQueryRow>(pageSize);
+
+            // Work out whether to use the response we've already got, or create a new one.
+            GetQueryResultsResponse response = _response;
+            if ((clonedOptions.PageSize == null || clonedOptions.PageSize > pageSize) && _response.Rows?.Count > pageSize)
```

Comment:
I'm unclear why this isn't just `if (_response.Rows?.Count > pageSize)`. If `_response` contains more than `pageSize` rows, doesn't it always need re-starting?

Reply:
Yes, I think you're right. Will simplify it in both places.